:sparkles: Add release notes expander functionality

chandankumar4 opened the pull request:
<!-- Thanks for sending a pull request! Here are some tips for you:
    1. If this is your first time, please read our contributor guidelines: https://github.com/kubernetes-sigs/cluster-api/blob/main/CONTRIBUTING.md#contributing-a-patch and developer guide https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/book/src/developer/guide.md

    2. Please add an icon to the title of this PR (see https://sigs.k8s.io/cluster-api/CONTRIBUTING.md#contributing-a-patch), and delete this line and similar ones
    the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🌱 (:seedling:, minor or other) 
-->

**What this PR does / why we need it**:

Add expander in generated release notes

```bash
./bin/notes --release v1.6.0 --pre-release-version true > CHANGELOG/v1.6.0.md


$ cat CHANGELOG/v1.6.0.md 
🚨 This is a RELEASE CANDIDATE. Use it only for testing purposes. If you find any bugs, file an [issue](https://github.com/kubernetes-sigs/cluster-api/issues/new).
## 👌 Kubernetes version support

- Management Cluster: v1.**X**.x -> v1.**X**.x
- Workload Cluster: v1.**X**.x -> v1.**X**.x

[More information about version support can be found here](https://cluster-api.sigs.k8s.io/reference/versions.html)

## Highlights

* REPLACE ME

## Deprecation Warning

REPLACE ME: A couple sentences describing the deprecation, including links to docs.

* [GitHub issue #REPLACE ME](REPLACE ME)

## Changes since v1.5.0
## :chart_with_upwards_trend: Overview
- 417 new commits merged
- 6 breaking changes :warning:
- 16 feature additions ✨
- 46 bugs fixed 🐛

## :memo: Proposals
- Community meeting: Add proposal for karpenter integration feature group (#9571)

<details>
<summary>More details about the release</summary>

## :warning: Breaking Changes
- API: Remove v1alpha3 API Version (#8997)
- API: Stop serving v1alpha4 API Versions (#8996)
- clusterctl: Improve Context handling in clusterctl (#8939)
- Dependency: Bump to controller-runtime v0.16 (#8999)
- MULTIPLE_AREAS[Metrics/Logging]: Implement secure diagnostics (metrics, pprof, log level changes) (#9264)
- util: : Remove go-vcs dependency from releaselink tool (#9288)

## :sparkles: New Features
- Testing: V1.29: Prepare quickstart, capd and tests for the new release including kind bump (#9890)
.....

## :bug: Bug Fixes

- util: Fix AddAnnotations for unstructured.Unstructured (#9164)
.....

## :seedling: Others
- API: Add ClusterClass column to Cluster CRD (#9120)
....

:book: Additionally, there have been 65 contributions to our documentation and book. (#10024, #10047, #8260, #8500, #8678, #8819, #8988, #9001, #9013, #9014, #9024, #9029, #9080, #9081, #9087, #9112, #9119, #9141, #9146, #9150, #9161, #9173, #9208, #9209, #9213, #9214, #9232, #9270, #9286, #9291, #9305, #9328, #9364, #9386, #9403, #9415, #9429, #9433, #9463, #9487, #9488, #9490, #9511, #9513, #9514, #9527, #9550, #9559, #9565, #9572, #9577, #9590, #9593, #9613, #9635, #9654, #9706, #9815, #9816, #9824, #9830, #9878, #9902, #9951, #9979) 


</details>
<br/>

_Thanks to all our contributors!_ 😊

```

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #9319 

<!-- 
Please label this pull request according to what area(s) you are addressing. For reference on PR/issue labels, see: https://github.com/kubernetes-sigs/cluster-api/labels?q=area+

Area example:
/area runtime-sdk
-->
/area release

willie-yao commented:
/ok-to-test

willie-yao reviewed (commented):
Nice work! I generated release notes for `v1.6.0-rc.1` and compared it to the existing release notes [here](https://github.com/kubernetes-sigs/cluster-api/releases/tag/v1.6.0-rc.1). It looks like the intention is to move everything under `Kubernetes version support` into the dropdown section, and to list changes from the previous release candidate or beta version (`v1.6.0-rc.0`) in this case outside of the dropdown. This is because the generated content is from v1.5.0 rather than the previous release candidate or beta version.  Let me know if this makes sense!

chandankumar4 commented:
> Nice work! I generated release notes for `v1.6.0-rc.1` and compared it to the existing release notes [here](https://github.com/kubernetes-sigs/cluster-api/releases/tag/v1.6.0-rc.1). It looks like the intention is to move everything under `Kubernetes version support` into the dropdown section, and to list changes from the previous release candidate or beta version (`v1.6.0-rc.0`) in this case outside of the dropdown. This is because the generated content is from v1.5.0 rather than the previous release candidate or beta version. Let me know if this makes sense!

Made the changes accordingly. Added `--previous-release-version` flag, so the example command will looks like
```
./bin/notes --release v1.6.0-rc.1 --previous-release-version tags/v1.6.0-rc.0  --pre-release-version true > CHANGELOG/v1.6.0-rc.1.md
```
And generated notes will be
```
🚨 This is a RELEASE CANDIDATE. Use it only for testing purposes. If you find any bugs, file an [issue](https://github.com/kubernetes-sigs/cluster-api/issues/new).
## Highlights

* REPLACE ME

## Deprecation Warning

REPLACE ME: A couple sentences describing the deprecation, including links to docs.

* [GitHub issue #REPLACE ME](REPLACE ME)

## Changes since v1.6.0-rc.0
## :chart_with_upwards_trend: Overview
- 57 new commits merged
- 2 feature additions ✨
- 14 bugs fixed 🐛

## :sparkles: New Features
- Control-plane: KCP: Allow mutation of all fields that should be mutable (#9885)
- Testing: V1.29: Prepare quickstart, capd and tests for the new release including kind bump (#9890)

## :bug: Bug Fixes
- CAPD: Fix ignition to also set the kube-proxy configuration to skip setting sysctls (#9895)
- ClusterCacheTracker: Fix ClusterCacheTracker memory leak (#10064)
- clusterctl: Validate no objects exist from CRDs before deleting them (#9834)
- ClusterResourceSet: Requeue after 1 minute if ErrClusterLocked got hit (#9788)
- e2e: Make MachinePools and MachineDeployments optional in ApplyClusterTemplateAndWait (#9962)
- e2e: Test: Fix finalizers test to not only rely on namespaced name (#10004)
- e2e: Test: retry GetOwnerGraph in owner references test on certificate errors (#10217)
- e2e: Test: wait for topology to get rolled out before continuing with scaling checks (#9827)
- IPAM: Fix webhooks using mixed api versions (#9863)
- Logging: Improve log k/v pairs and a improve/drop a few log lines (#9880)
- Machine: Watch external objects for machine before deleting (#10177)
- MachinePool: Fix TestReconcileMachinePoolScaleToFromZero flakes (#9822)
- MULTIPLE_AREAS[API/e2e]: Restore v1alpha3/v1alpha4 conversion to fix SSA issue & add e2e test coverage (#10151)
- Testing: Ssa: fix flaky test TestPatch/Test patch with Machine (#9915)

## :seedling: Others
- CABPK: Add pod metadata to capbk manager (#10212)
- CI: Bump kubebuilder envtest to 1.29.0 (#10014)
- clusterctl: Add 0 default to worker-machine-count help (#10205)
- clusterctl: Add hivelocity infra provider to clusterctl (#10168)
- clusterctl: Add hivelocity infra provider to clusterctl (#10222)
- clusterctl: Add Proxmox provider quickstart (#9812)
- clusterctl: Bump cert-manager to 1.14.2 (#10127)
- clusterctl: Bump cert-manager to v1.14.1 (#10120)
- clusterctl: Clarify rules for adding new clusterctl default providers (#10109)
- Community meeting: Promote chrischdi to Cluster API maintainer (#10089)
- Dependency: Bump controller runtime v0.16.5 (#10163)
- Dependency: Bump controller-runtime v0.16.4 (#10132)
- Dependency: Bump Go to 1.21.5 (#10152)
- Dependency: Bump golang.org/x/crypto to v0.17.0 (#9943)
- Dependency: Bump to Go 1.20.12 (#9841)
- Dependency: Go.mod: bump github.com/cloudflare/circl to v1.3.7 (#10002)
- e2e: Add DeepCopy method for E2EConfig (#9989)
- e2e: Add test util to resolve Kubernetes versions (#9996)
- e2e: Bump cluster-autoscaler to v1.29.0 (#9948)
- e2e: Drop duplicated scale test (#9976)
- e2e: Improve output of exec.KubectlApply (#9761)
- e2e: Make etcd and CoreDNS optional in upgrade and self-hosted tests (#9965)
- e2e: Test/framework add WatchDaemonSetLogsByLabelSelector method (#9994)
- e2e: Use manager in test extension (#10106)
- MachinePool: Use SSA Patch to create machines in MP controller (#9802)
- Release: Improve release-staging build (#10190)
- Release: Read in dockerfiles from stdin (#10209)
- Testing: Print conformance image used in kubetest (#10081)
- Testing: Test: add dynamic finalizer assertions for e2e framework (#10003)


🚨 This is a RELEASE CANDIDATE. Use it only for testing purposes. If you find any bugs, file an [issue](https://github.com/kubernetes-sigs/cluster-api/issues/new).
<details>
<summary>More details about the release</summary>
## 👌 Kubernetes version support

- Management Cluster: v1.**X**.x -> v1.**X**.x
- Workload Cluster: v1.**X**.x -> v1.**X**.x

[More information about version support can be found here](https://cluster-api.sigs.k8s.io/reference/versions.html)

## Highlights

* REPLACE ME

## Deprecation Warning

REPLACE ME: A couple sentences describing the deprecation, including links to docs.

* [GitHub issue #REPLACE ME](REPLACE ME)

## Changes since v1.5.0
## :chart_with_upwards_trend: Overview
- 435 new commits merged
- 6 breaking changes :warning:
- 16 feature additions ✨
- 49 bugs fixed 🐛

## :memo: Proposals
- Community meeting: Add proposal for karpenter integration feature group (#9571)

## :warning: Breaking Changes
- API: Remove v1alpha3 API Version (#8997)
- API: Stop serving v1alpha4 API Versions (#8996)
- clusterctl: Improve Context handling in clusterctl (#8939)
- Dependency: Bump to controller-runtime v0.16 (#8999)
- MULTIPLE_AREAS[Metrics/Logging]: Implement secure diagnostics (metrics, pprof, log level changes) (#9264)
- util: : Remove go-vcs dependency from releaselink tool (#9288)

## :sparkles: New Features
- API: Add validation to nested ObjectMeta fields (#8431)
- CAPD: Add config maps to CAPD RBAC (#9528)
- CAPD: Add MachinePool Machine implementation to CAPD components (#8842)
- CAPD: Allow adding custom HA proxy config for CAPD load balancer (#8785)
- CAPD: Initialize configmap object before getting it (#9529)
- ClusterClass: Introduce NamingStrategy and allow generating names using go templates (#9340)
- ClusterClass: Update API with ClusterClass MachinePool support (#8820)
- clusterctl: Add RKE2 bootstrap provider to clusterctl (#9720)
- clusterctl: Block move with annotation (#8690)
- Control-plane: KCP: Allow mutation of all fields that should be mutable (#9885)
- IPAM: Promote IPAM types to v1beta1 (#9525)
- MachinePool: Add MachinePool workers support in ClusterClass (#9016)
- MachineSet: Adjust preflight check to allow kubelet version skew of 3 for clusters running v1.28 and above (#9222)
- MULTIPLE_AREAS[Testing/Documentation]: V1.28: Prepare quickstart, capd and tests for the new release including kind bump (#9160)
- Release: Add automation to create release branch and tags (#9111)
- Testing: V1.29: Prepare quickstart, capd and tests for the new release including kind bump (#9890)

## :bug: Bug Fixes
- CABPK: Certificate paths in cloud-init scripts should not use a platform-dependent path separator (#9167)
- CAPD: Delete container after failed start to work around port allocation issues (#9125)
- CAPD: Fix DockerMachine panic (#9673)
- CAPD: Fix ignition to also set the kube-proxy configuration to skip setting sysctls (#9895)
- CI: Fix reporting bug in verify-container-image script (#9676)
- ClusterCacheTracker: Fix ClusterCacheTracker memory leak (#10064)
- ClusterClass: Add topology-owned label to MachineHealthChecks. (#9191)
- ClusterClass: Fix ClusterClass enqueue for ExtensionConfig (#9133)
- ClusterClass: Topology: fix namingstrategy webhook to not use uppercase characters for testing the template and align unit test to e2e test (#9425)
- clusterctl: Fix provider namespace secret not included in clusterctl move (#9694)
- clusterctl: Validate no objects exist from CRDs before deleting them (#9834)
- ClusterResourceSet: Requeue after 1 minute if ErrClusterLocked got hit (#9788)
- Dependency: Bump golang.org/x/net to v0.13.0 (#9121)
- Dependency: Bump to docker v24.0.5-0.20230714235725-36e9e796c6fc (#9038)
- Devtools: Adding metrics container port in tilt-prepare only if it's missing (#9308)
- Devtools: Allow duplicate objects in Tiltfile (#9302)
- Devtools: Change tilt debug base image to golang (#9070)
- Devtools: Fix tilt-prepare leader-elect setting (#9315)
- Devtools: Pin Plantuml version (#9424)
- Devtools: Tilt: ensure .tiltbuild/bin directory is created early enough, add tilt troubleshooting guide (#9165)
- e2e: Drop MachinePools from Dualstack tests (#9477)
- e2e: Fix autoscaler image repo (#9353)
- e2e: Make MachinePools and MachineDeployments optional in ApplyClusterTemplateAndWait (#9962)
- e2e: Test: Fix finalizers test to not only rely on namespaced name (#10004)
- e2e: Test: pin conformance image to a version which includes a fix for the dualstack tests (#9252)
- e2e: Test: retry GetOwnerGraph in owner references test on certificate errors (#10217)
- e2e: Test: wait for topology to get rolled out before continuing with scaling checks (#9827)
- IPAM: Fix webhooks using mixed api versions (#9863)
- KCP: Allow dropping patches KubeadmControlPlane KubeadmConfig (#9698)
- KCP: Allow unsetting useExperimentalRetryJoin field from KubeadmControlPlane (#9170)
- KCP: Fix KCP Controller reconcile always return error when workload cluster is unreachable (#9342)
- KCP: Requeue KCP object if ControlPlaneComponentsHealthyCondition is not yet true (#9032)
- Logging: Improve log k/v pairs and a improve/drop a few log lines (#9880)
- Machine: Retry Node delete when CCT is locked (#9570)
- Machine: Watch external objects for machine before deleting (#10177)
- MachineDeployment: MD controller: use regular random suffix for MachineSets, ensure max length 63 (#9298)
- MachineHealthCheck: Fix excessive trace logging in the machine health check controller (#9419)
- MachinePool: Fix bug where MachinePool Machine ownerRefs weren't updating (#9619)
- MachinePool: Fix TestReconcileMachinePoolScaleToFromZero flakes (#9822)
- MULTIPLE_AREAS[API/e2e]: Restore v1alpha3/v1alpha4 conversion to fix SSA issue & add e2e test coverage (#10151)
- Release: Alphabetically sorting release tool output (#9055)
- Release: Deduplicating area in pr title in release notes (#9186)
- Release: Fix go install path for kpromo v4.0.4 (#9336)
- Release: Generate warning when release notes can not be generated (#9163)
- Release: Hack/release-notes: ensure relase notes tool can be used for external projects again (#9018)
- Release: Reverts pull request from cahillsf/improve-release-speed (#9465)
- Runtime SDK: Set User Agent for test extension correctly (#9748)
- Testing: Ssa: fix flaky test TestPatch/Test patch with Machine (#9915)
- util: Fix AddAnnotations for unstructured.Unstructured (#9164)

## :seedling: Others
- API: Add ClusterClass column to Cluster CRD (#9120)
- API: Add verify-import-restrictions to enforce import restrictions (#9407)
- API: Enforce import restrictions in all API packages (#9461)
- API: Ensure we generate all webhook manifests.yaml (#9621)
- API: Move API v1beta1 webhooks to a separate package (#9047)
- API: Move docker infrastructure API v1beta1 webhooks to separate package (#9458)
- API: Move docker infrastructure experimental API v1beta1 webhooks to separate package (#9460)
- API: Move experimental addons API v1beta1 webhooks to separate package (#9438)
- API: Move experimental API v1beta1 webhooks to separate package (#9417)
- API: Move inmemory infrastructure API v1beta1 webhooks to separate package (#9459)
- API: Move Kubeadm API v1beta1 webhooks to separate package (#9410)
- API: Remove files and markers for Kubebuilder (#9344)
- API: Remove reliance on controller-runtime scheme builder (#9045)
- API: Remove reliance on controller-runtime scheme builder for experimental APIs (#9185)
- API: Remove reliance on controller-runtime scheme builder for remaining API groups (#9266)
- API: Remove the dependency on cluster-api/utils from addons API (#9482)
- API: Test and document controller ownerReferences (#9153)
- CABPK: Add pod metadata to capbk manager (#10212)
- CABPK: Remove Kubeadm upstream v1beta1 types (#9345)
- CAPD: Fix multi error handling in RunContainer (#9139)
- CAPD: Set Condition, if creating external LB failed. (#9697)
- CI: Add colored-line-number output for golangci-lint action (#9147)
- CI: Add dependabot for test and hack/tools module (#9041)
- CI: Add dependabot groups. Allow additional patch updates (#9263)
- CI: Add exclude for Kustomize API to dependabot config (#9059)
- CI: Add licence-scan for pull requests (#9184)
- CI: Add loggercheck linter and fix findings (#9446)
- CI: Add verify-govulncheck and verify-vulnerabilities targets and integrate to scan action (#9144)
- CI: Bump actions/cache from 3.3.1 to 3.3.2 (#9395)
- CI: Bump actions/checkout from 3.5.3 to 3.6.0 (#9324)
- CI: Bump actions/checkout from 3.6.0 to 4.0.0 (#9367)
- CI: Bump actions/checkout from 4.0.0 to 4.1.0 (#9497)
- CI: Bump actions/checkout from 4.1.0 to 4.1.1 (#9611)
- CI: Bump actions/github-script from 6.4.1 to 7.0.0 (#9713)
- CI: Bump actions/github-script from 7.0.0 to 7.0.1 (#9738)
- CI: Bump actions/setup-go from 4.0.1 to 4.1.0 (#9187)
- CI: Bump apidiff to v0.7.0 (#9472)
- CI: Bump golangci-lint to v1.54.1 (#9174)
- CI: Bump golangci/golangci-lint-action from 3.6.0 to 3.7.0 (#9261)
- CI: Bump kubebuilder envtest to 1.29.0 (#10014)
- CI: Bump tj-actions/changed-files from 37.5.0 to 37.6.1 (#9260)
- CI: Bump tj-actions/changed-files from 37.6.1 to 38.0.0 (#9276)
- CI: Bump tj-actions/changed-files from 38.0.0 to 38.1.3 (#9325)
- CI: Bump tj-actions/changed-files from 38.1.3 to 39.0.0 (#9368)
- CI: Bump tj-actions/changed-files from 39.0.0 to 39.0.1 (#9394)
- CI: Bump tj-actions/changed-files from 39.0.1 to 39.1.0 (#9454)
- CI: Bump tj-actions/changed-files from 39.1.0 to 39.2.0 (#9496)
- CI: Bump tj-actions/changed-files from 39.2.0 to 39.2.1 (#9535)
- CI: Bump tj-actions/changed-files from 39.2.1 to 39.2.2 (#9560)
- CI: Bump tj-actions/changed-files from 39.2.2 to 39.2.4 (#9610)
- CI: Bump tj-actions/changed-files from 39.2.4 to 40.0.0 (#9641)
- CI: Bump tj-actions/changed-files from 40.0.0 to 40.1.0 (#9678)
- CI: Bump tj-actions/changed-files from 40.1.0 to 40.1.1 (#9714)
- CI: Bump trivy to v0.45.0 (#9350)
- CI: Bump Trivy to v0.45.1 (#9445)
- CI: Fix .golangci.yml comments (#9499)
- CI: Ginkgolinter: forbid focus container (#9320)
- CI: Github: add edited and reopened as triggers for the GH workflow approval (#9259)
- CI: Github: add workflow to auto-approve golangci-lint if ok-to-test label is set (#9244)
- CI: Go.mod reformat to have only two require blocks (#9192)
- CI: Golangci-lint: replace deprecated local-prefixes setting for gci (#9339)
- CI: Make GO_ARCH explicit in verify_containter script (#9341)
- CI: Update actions for 1.5 and make names consistent (#9115)
- CI: Verify plantuml image generation in CI (#9363)
- ClusterCacheTracker: Add separate concurrency flag for cluster cache tracker (#9116)
- ClusterCacheTracker: Ensure Get/List calls are not getting stuck when apiserver is unreachable (#9028)
- ClusterCacheTracker: Fix accessor deletion on health check failure (#9025)
- ClusterClass: Add additional test cases for nested variable defaulting (#9728)
- ClusterClass: Add ownerRefs to BootstrapConfig/InfraMachinePool in classy Clusters (#9389)
- ClusterClass: Embed ssa.FilterObjectInput into HelperOption to remove duplication (#9512)
- ClusterClass: Fix some nits in Cluster topology engine tests (#9464)
- ClusterClass: Improve Cluster variable defaulting/validation errors (#9452)
- ClusterClass: Improve message for TopologyReconciledCondition (#9400)
- ClusterClass: Make ClusterClass generated object names consistent (#9254)
- ClusterClass: Minor fixes for CC+MP implementation (#9318)
- clusterctl: Add 0 default to worker-machine-count help (#10205)
- clusterctl: Add hivelocity infra provider to clusterctl (#10168)
- clusterctl: Add hivelocity infra provider to clusterctl (#10222)
- clusterctl: Add Proxmox provider quickstart (#9812)
- clusterctl: Bump cert-manager to 1.14.2 (#10127)
- clusterctl: Bump cert-manager to v1.14.1 (#10120)
- clusterctl: Check resource blocking clusterctl move during discovery (#9246)
- clusterctl: Clarify rules for adding new clusterctl default providers (#10109)
- clusterctl: Propagate ctx to retryWithExponentialBackoff in clusterctl (#9437)
- clusterctl: Use goproxy to check version in clusterctl (#9237)
- clusterctl: Use http get to download files from GitHub in clusterctl (#9236)
- Community meeting: Promote chrischdi to Cluster API maintainer (#10089)
- Dependency: Bump cert-manager to v1.12.4 (#9358)
- Dependency: Bump cert-manager to v1.13.0 (#9408)
- Dependency: Bump cert-manager to v1.13.1 (#9505)
- Dependency: Bump cert-manager to v1.13.2 (#9653)
- Dependency: Bump cloud.google.com/go/storage from 1.29.0 to 1.31.0 in /hack/tools (#9053)
- Dependency: Bump cloud.google.com/go/storage from 1.31.0 to 1.32.0 in /hack/tools (#9218)
- Dependency: Bump cloud.google.com/go/storage from 1.32.0 to 1.33.0 in /hack/tools (#9423)
- Dependency: Bump cloud.google.com/go/storage from 1.33.0 to 1.34.1 in /hack/tools (#9661)
- Dependency: Bump cloud.google.com/go/storage from 1.34.1 to 1.35.1 in /hack/tools (#9727)
- Dependency: Bump controller runtime v0.16.5 (#10163)
- Dependency: Bump controller tools to v0.13.0 (#9221)
- Dependency: Bump controller-runtime to v0.15.1 (#9126)
- Dependency: Bump controller-runtime to v0.16.3 (#9592)
- Dependency: Bump controller-runtime v0.16.4 (#10132)
- Dependency: Bump conversion-gen to v0.28 (#9267)
- Dependency: Bump corefile-migration library to v1.0.21 (#9307)
- Dependency: Bump docker to v24.0.5 (#9064)
- Dependency: Bump envtest binaries to 1.28 (#9268)
- Dependency: Bump github.com/blang/semver to v4 (#9189)
- Dependency: Bump github.com/docker/distribution (#9544)
- Dependency: Bump github.com/docker/docker from 24.0.5+incompatible to 24.0.6+incompatible in /test (#9377)
- Dependency: Bump github.com/docker/docker from 24.0.6+incompatible to 24.0.7+incompatible in /test (#9652)
- Dependency: Bump github.com/emicklei/go-restful/v3  from 3.9.0 to 3.10.2 in /test (#9048)
- Dependency: Bump github.com/emicklei/go-restful/v3 from 3.10.2 to 3.11.0 in /test (#9272)
- Dependency: Bump github.com/evanphx/json-patch/v5 from 5.6.0 to 5.7.0 (#9397)
- Dependency: Bump github.com/fatih/color from 1.15.0 to 1.16.0 (#9681)
- Dependency: Bump github.com/go-logr/logr from 1.2.4 to 1.3.0 (#9644)
- Dependency: Bump github.com/google/go-cmp from 0.5.9 to 0.6.0 (#9562)
- Dependency: Bump github.com/google/go-cmp from 0.5.9 to 0.6.0 in /test (#9538)
- Dependency: Bump github.com/onsi/ginkgo/v2 from 2.11.0 to 2.12.0 (#9327)
- Dependency: Bump github.com/onsi/ginkgo/v2 from 2.12.0 to 2.12.1 (#9498)
- Dependency: Bump github.com/onsi/ginkgo/v2 from 2.12.0 to 2.12.1 in /test (#9467)
- Dependency: Bump github.com/onsi/ginkgo/v2 from 2.12.1 to 2.13.0 (#9533)
- Dependency: Bump github.com/onsi/ginkgo/v2 from 2.12.1 to 2.13.0 in /test (#9539)
- Dependency: Bump github.com/onsi/ginkgo/v2 from 2.13.0 to 2.13.1 (#9716)
- Dependency: Bump github.com/onsi/gomega from 1.27.10 to 1.28.0 (#9516)
- Dependency: Bump github.com/onsi/gomega from 1.27.10 to 1.28.0 in /test (#9519)
- Dependency: Bump github.com/onsi/gomega from 1.27.8 to 1.27.10 (#9062)
- Dependency: Bump github.com/onsi/gomega from 1.28.0 to 1.28.1 (#9608)
- Dependency: Bump github.com/onsi/gomega from 1.28.1 to 1.29.0 (#9643)
- Dependency: Bump github.com/onsi/gomega from 1.29.0 to 1.30.0 (#9717)
- Dependency: Bump github.com/prometheus/client_golang from 1.16.0 to 1.17.0 (#9517)
- Dependency: Bump github.com/spf13/cobra from 1.7.0 to 1.8.0 (#9679)
- Dependency: Bump github.com/spf13/viper from 1.16.0 to 1.17.0 (#9561)
- Dependency: Bump Go to 1.21.5 (#10152)
- Dependency: Bump Go to v1.20.10 (#9551)
- Dependency: Bump Go version to v1.20.11 (#9683)
- Dependency: Bump go-github dependency to version v53 (#8995)
- Dependency: Bump go.etcd.io/etcd/api/v3 from 3.5.9 to 3.5.10 (#9642)
- Dependency: Bump go.etcd.io/etcd/client/v3 from 3.5.9 to 3.5.10 (#9646)
- Dependency: Bump go.opentelemetry.io/* dependencies (#9598)
- Dependency: Bump golang.org/grpc to v1.59.0 (#9626)
- Dependency: Bump golang.org/x/crypto to v0.17.0 (#9943)
- Dependency: Bump golang.org/x/net from 0.14.0 to 0.15.0 in /test (#9378)
- Dependency: Bump golang.org/x/net from 0.15.0 to 0.17.0 in /test (#9537)
- Dependency: Bump golang.org/x/oauth2 from 0.10.0 to 0.11.0 (#9142)
- Dependency: Bump golang.org/x/oauth2 from 0.11.0 to 0.12.0 (#9396)
- Dependency: Bump golang.org/x/oauth2 from 0.12.0 to 0.13.0 (#9534)
- Dependency: Bump golang.org/x/oauth2 from 0.13.0 to 0.14.0 (#9715)
- Dependency: Bump golang.org/x/text from 0.12.0 to 0.13.0 (#9370)
- Dependency: Bump golang.org/x/text from 0.13.0 to 0.14.0 (#9680)
- Dependency: Bump golangci-lint to v1.55.1 (#9622)
- Dependency: Bump golangci-lint to v1.55.2 (#9740)
- Dependency: Bump gomodules.xyz/jsonpatch/v2 from 2.3.0 to 2.4.0 (#9188)
- Dependency: Bump google.golang.org/api from 0.126.0 to 0.132.0 in /hack/tools (#9051)
- Dependency: Bump google.golang.org/api from 0.132.0 to 0.134.0 in /hack/tools (#9083)
- Dependency: Bump google.golang.org/api from 0.134.0 to 0.136.0 in /hack/tools (#9159)
- Dependency: Bump google.golang.org/api from 0.136.0 to 0.137.0 in /hack/tools (#9217)
- Dependency: Bump google.golang.org/api from 0.137.0 to 0.138.0 in /hack/tools (#9280)
- Dependency: Bump google.golang.org/api from 0.138.0 to 0.140.0 in /hack/tools (#9422)
- Dependency: Bump google.golang.org/api from 0.140.0 to 0.142.0 in /hack/tools (#9475)
- Dependency: Bump google.golang.org/api from 0.142.0 to 0.143.0 in /hack/tools (#9509)
- Dependency: Bump google.golang.org/api from 0.143.0 to 0.144.0 in /hack/tools (#9526)
- Dependency: Bump google.golang.org/api from 0.144.0 to 0.146.0 in /hack/tools (#9541)
- Dependency: Bump google.golang.org/api from 0.146.0 to 0.148.0 in /hack/tools (#9581)
- Dependency: Bump google.golang.org/api from 0.148.0 to 0.149.0 in /hack/tools (#9660)
- Dependency: Bump google.golang.org/api from 0.149.0 to 0.150.0 in /hack/tools (#9691)
- Dependency: Bump google.golang.org/grpc from 1.55.0 to 1.55.1 (#9284)
- Dependency: Bump google.golang.org/grpc from 1.58.2 to 1.58.3 (#9607)
- Dependency: Bump google.golang.org/grpc from 1.58.2 to 1.58.3 in /test (#9564)
- Dependency: Bump kpromo to v4.0.4 (#9241)
- Dependency: Bump opentelemetry to fix CVEs (#9709)
- Dependency: Bump sigs.k8s.io/controller-runtime from 0.16.0 to 0.16.1 (#9369)
- Dependency: Bump sigs.k8s.io/controller-runtime from 0.16.1 to 0.16.2 (#9456)
- Dependency: Bump sigs.k8s.io/controller-runtime from 0.16.1 to 0.16.2 in /hack/tools (#9421)
- Dependency: Bump sigs.k8s.io/controller-runtime from 0.16.1 to 0.16.2 in /test (#9406)
- Dependency: Bump sigs.k8s.io/controller-runtime from 0.16.2 to 0.16.3 in /hack/tools (#9576)
- Dependency: Bump sigs.k8s.io/controller-tools from 0.12.0 to 0.12.1 in /hack/tools (#9050)
- Dependency: Bump sigs.k8s.io/controller-tools from 0.12.1 to 0.13.0 in /hack/tools (#9281)
- Dependency: Bump sigs.k8s.io/yaml from 1.3.0 to 1.4.0 (#9645)
- Dependency: Bump sigs.k8s.io/yaml from 1.3.0 to 1.4.0 in /test (#9615)
- Dependency: Bump some dependencies in Makefile (#9549)
- Dependency: Bump the kubernetes group in /hack/tools with 2 updates (#9420)
- Dependency: Bump the kubernetes group in /hack/tools with 2 updates (#9574)
- Dependency: Bump the kubernetes group with 2 updates (#9606)
- Dependency: Bump the kubernetes group with 4 updates (#9326)
- Dependency: Bump the kubernetes group with 4 updates (#9455)
- Dependency: Bump the kubernetes group with 4 updates (#9739)
- Dependency: Bump to Go 1.20.12 (#9841)
- Dependency: Bump to Go 1.20.6 (#9057)
- Dependency: Bump to Go 1.20.7 (#9103)
- Dependency: Bump to Go 1.20.8 (#9381)
- Dependency: Bump trivy to v0.46.0 (#9558)
- Dependency: Bump Trivy to v0.47.0 (#9671)
- Dependency: Go.mod: bump github.com/cloudflare/circl to v1.3.7 (#10002)
- Dependency: Replace hashicorp/go-multierror with kerrors (#9175)
- Dependency: Update cert-manager to v1.12.3 (#9197)
- Dependency: Update ensure-kubectl.sh to 1.28 (#9275)
- Devtools: Add .PHONY for doctoc target (#9148)
- Devtools: Allow non-provider Deployments in Tilt (#9404)
- Devtools: Bump CAPI visualizer to v1.2.0 (#9195)
- Devtools: Drop duplicate pprof and unused linter excludes (#9156)
- Devtools: Improve Performance dashboard (#9387)
- Devtools: Make dev cluster networking configurable (#9183)
- Devtools: Makefile: run doctoc only once (#9182)
- Devtools: Move tilt-prepare and log-push to tools/internal (#9020)
- Devtools: Observability: move metrics to config and use sidecar in kube-state-metrics (#9390)
- Devtools: Refactor docker-push* Makefile targets so users can control with ALL_DOCKER_BUILD which images are pushed (#8586)
- Devtools: Tiltfile: rename deploy_kustomizations to additional_kustomizations (#9439)
- e2e: Add back flavor to Waiter interface (#9166)
- e2e: Add CRS re-reconcile to ownerReference test (#9296)
- e2e: Add DeepCopy method for E2EConfig (#9989)
- e2e: Add log level for kube components patch to ClusterClass (#9493)
- e2e: Add MachinePools to Topology Quickstart E2E Templates (#9393)
- e2e: Add MP timeouts to cluster upgrade tests (#9723)
- e2e: Add test for ownerReference apiVersion update (#9269)
- e2e: Add test for scale testing machinery (#9510)
- e2e: Add test util to resolve Kubernetes versions (#9996)
- e2e: Bump autoscaler to v1.28.0 (#9349)
- e2e: Bump cluster-autoscaler to v1.29.0 (#9948)
- e2e: Drop duplicated scale test (#9976)
- e2e: Drop PR-Informing test tag and job description (#9362)
- e2e: Dump all pods in e2e test clusters (#9441)
- e2e: Dump all resource information for self-hosted tests (#9547)
- e2e: Ensure finalizers are resilient on reconciliation (#9471)
- e2e: Fail tests if test env version check fails (#9388)
- e2e: Fix broken e2e test clusterclass (#9506)
- e2e: Improve labels/annotations in CAPD test ClusterClass (#9469)
- e2e: Improve logging for self-hosted e2e test (#9637)
- e2e: Improve output of exec.KubectlApply (#9761)
- e2e: Make etcd and CoreDNS optional in upgrade and self-hosted tests (#9965)
- e2e: Refactor e2e ownerRef test utils (#9313)
- e2e: Test/conformance: add variables to overwrite parralelization of kubetest runs (#9667)
- e2e: Test/e2e: structure resources by namespace/kind again (#9462)
- e2e: Test/framework add WatchDaemonSetLogsByLabelSelector method (#9994)
- e2e: Use existing value of `SKIP_RESOURCE_CLEANUP` if set in environment (#9152)
- e2e: Use manager in test extension (#10106)
- IPAM: Add age also to v1beta1 IPAM types (#9729)
- IPAM: Add age column to kubectl output (#9521)
- IPAM: Enforce IPAM import alias (#9730)
- KCP: Controlplane: add a test case for syncMachines where the InfraMachine does not exist. (#8992)
- KCP: Remove disableInPlacePropagation field in KCP controller (#9099)
- KCP: Remove redundant GetRESTConfig in KCP Management.GetWorkloadCluster (#9448)
- KCP: Support admin config for Kubeadm v1.29 (#9682)
- Logging: Change default log level to 2 (#9093)
- Logging: Fix patch errors not being logged (#9224)
- Logging: Set controller-runtime logger for clusterctl (#9107)
- MachinePool: Add MachinePool Builders (#9346)
- MachinePool: Add util function to get MachinePool by label (#9219)
- MachinePool: Use SSA Patch to create machines in MP controller (#9802)
- Metrics: Hack/observability: add capi_machine_status_certificatesexpirydate metric (#9084)
- Metrics: Hack: bump kube-state-metrics and prometheus charts (#9352)
- MISSING_AREA: Add CABPOCNE and CACPOCNE Providers (#9012)
- Release: Add additional blocks to release note generation (#9247)
- Release: Add adilGhaffarDev to release-team owner alias (#9718)
- Release: Add Release Team OWNERS file to docs/release folder (#9294)
- Release: Avoid cd'ing into test folders when building Docker images (#9744)
- Release: Bump kubebuilder tools to v0.4.2 (#9665)
- Release: Bump PR title checker github action to v0.4.1 (#9650)
- Release: Capitalize area prefix in release notes by default (#9614)
- Release: Capitalized title in release notes (#9086)
- Release: Clarify release team vs k8s/k8s-SIGs org membership (#9089)
- Release: Exclude release trigger PRs from release notes (#9444)
- Release: Format Community meeting area in release notes (#9648)
- Release: Format MachineHealthCheck area in release notes (#9500)
- Release: Give permissions for release notes tool to release team (#9563)
- Release: Hack: add weekly update script for Slack (#9343)
- Release: Improve multiple areas PRs with user friendly subs (#9071)
- Release: Improve release notes formatting (#9337)
- Release: Improve release speed (#9392)
- Release: Improve release staging build speed (#9536)
- Release: Improve release-staging build (#10190)
- Release: Prepare main branch for v1.6 development (#9097)
- Release: Read in dockerfiles from stdin (#10209)
- Release: Remove auto-labelling for clusterctl (#8990)
- Release: Remove extra separator after title in release notes (#9605)
- Release: Revert "Improve release staging build speed" (#9753)
- Release: Update instructions checklist to generate release notes (#9443)
- Release: Update kubebuilder-release-tools to v0.4.0 (#9531)
- Release: Update release-notes make target + corresponding doc (#9573)
- Runtime SDK: Improve ClusterClass watch for ExtensionConfigs (#9338)
- Testing: Add MachinePool test cases (#9474)
- Testing: Add MachinePool test cases to engine tests (#9373)
- Testing: Add MachinePool test cases to variables tests (#9372)
- Testing: Add MachinePools to Runtime SDK and Rollout tests (#9703)
- Testing: Add MachinePools to topology upgrade test (#9502)
- Testing: Add test for required properties in clusterclass variables (#9113)
- Testing: Add unit tests for CC MP blueprint, current_state, & desired_state (#9348)
- Testing: Add unit tests for reconcile_state, cluster_controller, & conditions (#9380)
- Testing: Add WaitForMachinePoolToBeUpgraded to self-hosted test (#9540)
- Testing: Extend test/framework to collect workload cluster nodes (#9416)
- Testing: Follow-up fixes: Add MachinePools to Runtime SDK and Rollout tests (#9719)
- Testing: Print conformance image used in kubetest (#10081)
- Testing: Replacing gomega.Equal with gomega.BeComparableTo (#9015)
- Testing: Test: add dynamic finalizer assertions for e2e framework (#10003)
- util: Adjust naming in SortForCreate implementation (#9311)
- util: Move `internal.labels` to `format` package for use by providers (#9002)
- util: Refactor SortForCreate to use sort.Slice (#9251)
- util: Remove previously deprecated code (#9136)

:book: Additionally, there have been 67 contributions to our documentation and book. (#10024, #10047, #10105, #10116, #8260, #8500, #8678, #8819, #8988, #9001, #9013, #9014, #9024, #9029, #9080, #9081, #9087, #9112, #9119, #9141, #9146, #9150, #9161, #9173, #9208, #9209, #9213, #9214, #9232, #9270, #9286, #9291, #9305, #9328, #9364, #9386, #9403, #9415, #9429, #9433, #9463, #9487, #9488, #9490, #9511, #9513, #9514, #9527, #9550, #9559, #9565, #9572, #9577, #9590, #9593, #9613, #9635, #9654, #9706, #9815, #9816, #9824, #9830, #9878, #9902, #9951, #9979) 


</details>
<br/>
_Thanks to all our contributors!_ 😊

```

willie-yao reviewed (commented):
I ran the following command as it was listed in the PR description and get a a panic: `panic: runtime error: index out of range [1] with length 1`
```
./bin/notes --release v1.6.0 --pre-release-version true > CHANGELOG/v1.6.0.md
```

I think you might need to enforce setting `--previous-release-version` if `--pre-release-version` is set to true.

willie-yao commented on hack/tools/release/notes/print.go:
```diff
@@ -79,7 +81,11 @@ func (p *releaseNotesPrinter) print(entries []notesEntry) {
 		fmt.Printf("🚨 This is a RELEASE CANDIDATE. Use it only for testing purposes. If you find any bugs, file an [issue](https://github.com/%s/issues/new).\n", p.repo)
```
This could either be a release candidate or beta release. You can either change this based on how the release tag string is formatted, or just include both `RELEASE CANDIDATE/BETA RELEASE` and have the user manually change this.

willie-yao commented:
Looks like there are still some errors when running the release notes tool test with `make test-release-notes-tool`. I think it's just some minor formatting that you need to fix:
```
 -       <details>
        -       <summary>More details about the release</summary>
                ## 👌 Kubernetes version support
          
                ... // 188 identical lines
                - google.golang.org/grpc/cmd/protoc-gen-go-grpc: v1.1.0
          
        -       </details>
        -       <br/>
 ```

willie-yao reviewed (commented):
Thanks for making this change! I think the unit test `make test-release-notes-tool` is still failing.

Also one more nit: The tool should be printing this line inside the dropdown:
https://github.com/willie-yao/cluster-api/blob/main/CHANGELOG/v1.6.0-rc.1.md#L32

fabriziopandini commented:
@willie-yao @cahillsf PTAL when you have time (I will approve after your lgtm)

willie-yao reviewed (commented):
Sorry for not catching this earlier, but the `Kubernetes Version Support` section shouldn't be printed outside of the dropdown for a pre-release version as seen here: https://github.com/willie-yao/cluster-api/blob/main/CHANGELOG/v1.7.0-beta.0.md

Once that is fixed I will lgtm!

willie-yao reviewed (approved):
/lgtm
/hold for squash

Thanks for all your patience and hard work on this! 🚀

chandankumar4 commented:
/assign @fabriziopandini

willie-yao commented:
FYI #10324 may merge soon with a change to the flags for the release notes tool. You may need to rebase if this merges first. It just removes the `--pre-release-version` in favor of `--release-type` to distinguish between beta releases and release candidates.

tobiasgiese commented:
> removes the `--pre-release-version` in favor of `--release-type` to distinguish between beta releases and release candidates

That is not completely true. I removed the parameter `--pre-release-version` completely and let it determine from the given tag what type of release it is :)

willie-yao commented:
I tried running this command `./bin/notes --release v1.6.0-rc.1 --previous-release-version tags/v1.6.0-rc.0 > CHANGELOG/v1.6.0-rc.1.md` and the dropdown seems to be missing. You might need to adjust when to add the dropdown based on if the release tag is a preview version or not, now that there is no longer a `--pre-release-version` flag

chandankumar4 commented:
@willie-yao  sorry for the confusion, It messed up while resolving review comment as `--pre-release-version` flag is removed. I have fixed it now. Thanks

willie-yao commented:
No problem! Thanks for the rebase and quick fix!

/lgtm

tobiasgiese commented on hack/tools/release/notes/print.go:
```diff
@@ -75,11 +78,23 @@ func (p *releaseNotesPrinter) print(entries []notesEntry, commitsInRelease int,
+		fmt.Print(`<details>
+<summary>More details about the release</summary>
+
+:warning: **RELEASE CANDIDATE NOTES** :warning:
```
Should we change here between beta and rc release?

chandankumar4 replied:
It's for both, thanks for pointing out, made the changes.

tobiasgiese commented on hack/tools/release/notes/process.go:
```diff
@@ -117,10 +117,18 @@ func (g prEntriesProcessor) process(prs []pr) []notesEntry {
+func (d dependenciesProcessor) generateDependencies(previousRelease ref) (string, error) {
 	repoURL := fmt.Sprintf("https://github.com/%s", d.repo)
+
+	var fromTag string
+	if previousRelease.value != "" {
+		fromTag = previousRelease.value
+	} else {
+		fromTag = d.fromTag
+	}
```
```suggestion
	fromTag = d.fromTag
	if previousRelease.value != "" {
		fromTag = previousRelease.value
	}
```

tobiasgiese commented:
/lgtm

fabriziopandini commented:
/lgtm
/approve

Let's start using this and eventually iterate while working on 1.7 pre releases

chandankumar4 commented:
/unhold
